pal/src/host/linux-sgx/utils/fast_clock.c

```diff
@@ -0,0 +1,362 @@
+}
+
+static int handle_state_init(fast_clock_t* fast_clock, fast_clock_desc_t descriptor, uint64_t* time_usec);
+static int handle_state_calibrating(fast_clock_t* fast_clock, fast_clock_desc_t descriptor, uint64_t* time_usec);
+static inline int handle_state_rdtsc(fast_clock_t* fast_clock, fast_clock_desc_t descriptor, uint64_t* time_usec, bool force_new_timepoint);
+static inline int handle_state_rdtsc_recalibrate(fast_clock_t* fast_clock, fast_clock_desc_t descriptor, uint64_t* time_usec);
+static int handle_state_rdtsc_disabled(uint64_t* time_usec);
+
+int fast_clock__get_time(fast_clock_t* fast_clock, uint64_t* time_usec, bool force_new_timepoint)
+{
+    fast_clock_desc_t descriptor = desc_atomic_load(fast_clock, __ATOMIC_ACQUIRE);
+    switch (descriptor.state)
+    {
+    case FC_STATE_RDTSC:
+        return handle_state_rdtsc(fast_clock, descriptor, time_usec, force_new_timepoint);
+    case FC_STATE_RDTSC_RECALIBRATE:
+        return handle_state_rdtsc_recalibrate(fast_clock, descriptor, time_usec);
+    case FC_STATE_CALIBRATING:
+        return handle_state_calibrating(fast_clock, descriptor, time_usec);
+    case FC_STATE_INIT:
+        return handle_state_init(fast_clock, descriptor, time_usec);
+    case FC_STATE_RDTSC_DISABLED:
+    default:
+        return handle_state_rdtsc_disabled(time_usec);
+    }
+}
+
+static inline bool set_change_state_guard(fast_clock_t* fast_clock, fast_clock_desc_t descriptor)
+{
+    if ((descriptor.flags & FC_FLAGS_STATE_CHANGING) != 0) {
+        return false;
+    }
+
+    fast_clock_desc_t state_change_guard_desc = {
+        .state = descriptor.state,
+        .flags = descriptor.flags | FC_FLAGS_STATE_CHANGING,
+    };
+
+    return __atomic_compare_exchange_n(
+        &fast_clock->atomic_descriptor.desc, &descriptor.desc, state_change_guard_desc.desc,
+        /*weak=*/false, __ATOMIC_RELAXED, __ATOMIC_RELAXED
+    );
+}
+
+bool fast_clock__is_enabled(const fast_clock_t* fast_clock)
+{
+    fast_clock_desc_t descriptor = desc_atomic_load(fast_clock, __ATOMIC_RELAXED);
+    return (descriptor.state != FC_STATE_RDTSC_DISABLED);
+}
+
+void fast_clock__disable(fast_clock_t* fast_clock)
+{
+    /* We need to busy-loop until the state change guard is acquired here - since fast-clock
+     * might be in the midst of transitioning states. We can't simply store the DISABLED state. */
+    fast_clock_desc_t descriptor;
+    do {
+        descriptor = desc_atomic_load(fast_clock, __ATOMIC_ACQUIRE);
+    } while(!set_change_state_guard(fast_clock, descriptor));
+
+    fast_clock_desc_t disabled_desc = advance_state(descriptor, FC_STATE_RDTSC_DISABLED, false);
+    desc_atomic_store(fast_clock, disabled_desc, __ATOMIC_RELEASE);
+}
+
+static inline fast_clock_timepoint_t* get_timepoint(fast_clock_t* fast_clock, fast_clock_desc_t descriptor)
+{
+    return &fast_clock->time_points[timepoint_index(descriptor)];
+}
+
+static bool is_rdtsc_available(void) {
```

Comment:
Not sure if this method is needed - can we rely on the RDTSC emulation feature?

Maybe we can do the following - have a global `g_is_rdtsc_emulated` which is set under the `FIRST_TIME()` guard in the emulator. This method will simply call `get_tsc()` then check the global. We can then completely remove this logic from `pal_sgx_main`.